REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [5]:
import pandas as pd
import numpy as np

In [6]:
bus = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('data/inspections.csv')

In [7]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [8]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [9]:
#6315 records
bus.count()

business_id     6315
name            6315
address         6315
city            6315
state           6315
postal_code     6056
latitude        3599
longitude       3599
phone_number    5696
dtype: int64

#### Question 2b: How many unique business IDs are there?  

In [10]:
ls = bus['business_id'].unique()
len(ls)

6315

#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [11]:
bus['name'].value_counts()

STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
WALGREENS                     11
SAN FRANCISCO SOUP COMPANY    11
                              ..
ABC SUPER MARKET INC           1
MR. T CAFE                     1
BREW BROTHERS CAFE, INC.       1
HAIGHT STREET MARKET           1
DUC LOI SUPERMARKET            1
Name: name, Length: 5692, dtype: int64

## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [12]:
#zip codes are stored as an object

bus.dtypes

business_id       int64
name             object
address          object
city             object
state            object
postal_code      object
latitude        float64
longitude       float64
phone_number     object
dtype: object

#### Question 3b

What are the unique values of postal_code?

In [13]:
bus['postal_code'].unique()

array(['94104', '94109', '94133', '94110', '94122', '94103', '94115',
       '94131', '94111', '94117', '94107', '94108', '94102', '94132',
       '94105', '94134', '94124', '94116', '94121', '94118', '94112',
       '94123', '94114', '94127', '941102019', '941', '94545', '94066',
       '941033148', nan, '94158', '95105', '94013', '94130', 'CA',
       '92672', '94120', '94143', '94609', '94101', '00000', '94188',
       '94621', '94014', '94129', '94602', 'Ca'], dtype=object)

#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [14]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [15]:
vb = bus[bus['postal_code'].isin(validZip)]

vb.count()

business_id     5999
name            5999
address         5999
city            5999
state           5999
postal_code     5999
latitude        3516
longitude       3516
phone_number    5444
dtype: int64

In [16]:
vb.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [17]:
vb[pd.isnull(vb['longitude'])].count()



business_id     2483
name            2483
address         2483
city            2483
state           2483
postal_code     2483
latitude           0
longitude          0
phone_number    2160
dtype: int64

#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [18]:
vb['null_lon'] = np.where(vb['longitude'].isnull(), 1, 0)
vb['not_null_lon'] = np.where(vb['longitude'].notna(), 1 , 0)

valid_zip = vb[['postal_code', 'null_lon', 'not_null_lon']].groupby('postal_code').sum()

valid_zip


<ipython-input-18-0068777fd960>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  vb['null_lon'] = np.where(vb['longitude'].isnull(), 1, 0)
<ipython-input-18-0068777fd960>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  vb['not_null_lon'] = np.where(vb['longitude'].notna(), 1 , 0)


null_lon  not_null_lon
postal_code                        
94102             196           262
94103             252           306
94104              59            74
94105             100           126
94107             247           204
94108              82           142
94109             146           237
94110             251           334
94111             115           167
94112              59           131
94114             108           122
94115              76           162
94116              35            58
94117              67           130
94118             100           143
94121              55           106
94122             123           147
94123              53           116
94124              88            76
94127              28            43
94131              14            34
94132              65            69
94133             135           282
94134              29            45

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [19]:
#do group percentages

valid_zip['pct_null_lon'] = valid_zip['null_lon'] / (valid_zip['null_lon'] + valid_zip['not_null_lon']) * 100

valid_zip.sort_values(['pct_null_lon'], ascending = False).head()


null_lon  not_null_lon  pct_null_lon
postal_code                                      
94107             247           204     54.767184
94124              88            76     53.658537
94132              65            69     48.507463
94114             108           122     46.956522
94122             123           147     45.555556

# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [34]:
ins.count()


business_id    15430
score          15430
date           15430
type           15430
dtype: int64

In [35]:
len(ins['business_id'].unique())

#shows that they are not unique for each busines

5730

#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [40]:
ins['type'].value_counts()

routine      15429
complaint        1
Name: type, dtype: int64

In [164]:
ins.groupby('type')['type'].transform()

type
complaint        1
routine      15429
Name: type, dtype: int64

In [51]:
ins[ins.groupby('type')['type'].transform('count').ge(10)]

0        True
1        True
2        True
3        True
4        True
         ... 
15425    True
15426    True
15427    True
15428    True
15429    True
Name: type, Length: 15430, dtype: bool

In [68]:
ins2 = ins[ins.groupby('type')['type'].transform('count').ge(10)]

ins2
           

business_id  score      date     type
0               10     82  20160503  routine
1               10     94  20140729  routine
2               10     92  20140114  routine
3               19     94  20160513  routine
4               19     94  20141110  routine
...            ...    ...       ...      ...
15425        89515     98  20161103  routine
15426        89701    100  20161206  routine
15427        89790     96  20161129  routine
15428        89803    100  20161215  routine
15429        89900     93  20161206  routine

[15429 rows x 4 columns]

#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [105]:
ins2['year']= ins2['date'].astype(str).str[0:4].astype(int)

ins2

<ipython-input-105-532297ee9b36>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ins2['year']= ins2['date'].astype(str).str[0:4].astype(int)


business_id  score      date     type  year
0               10     82  20160503  routine  2016
1               10     94  20140729  routine  2014
2               10     92  20140114  routine  2014
3               19     94  20160513  routine  2016
4               19     94  20141110  routine  2014
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[15429 rows x 5 columns]

#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [106]:
ins2.groupby(['year', 'type'])['type'].count()

year  type   
2013  routine      38
2014  routine    5629
2015  routine    3923
2016  routine    5839
Name: type, dtype: int64

In [118]:
ins3 = ins2[ins2.groupby('year')['year'].transform('count').ge(40)]

ins3

business_id  score      date     type  year
0               10     82  20160503  routine  2016
1               10     94  20140729  routine  2014
2               10     92  20140114  routine  2014
3               19     94  20160513  routine  2016
4               19     94  20141110  routine  2014
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[15391 rows x 5 columns]

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [122]:
ins4 = ins3[ins3['year'] == 2016]
ins4

business_id  score      date     type  year
0               10     82  20160503  routine  2016
3               19     94  20160513  routine  2016
6               24     98  20161005  routine  2016
7               24     96  20160311  routine  2016
10              31     84  20160816  routine  2016
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[5839 rows x 5 columns]

### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [127]:
full_table = pd.merge(bus, ins3, on="business_id", how="left")

full_table.head(10)

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           10                    TIRAMISU KITCHEN   
2           10                    TIRAMISU KITCHEN   
3           19               NRGIZE LIFESTYLE CAFE   
4           19               NRGIZE LIFESTYLE CAFE   
5           19               NRGIZE LIFESTYLE CAFE   
6           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
7           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
8           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
9           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
2                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
3   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
4   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
5   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
6  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
7  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
8  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
9  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   

    longitude  phone_number  score        date     type    year  
0 -122.403816  +14154217044   82.0  20160503.0  routine  2016.0  
1 -122.403816  +14154217044   94.0  20140729.0  routine  2014.0  
2 -122.403816  +14154217044   92.0  20140114.0  routine  2014.0  
3 -122.421547  +14157763262   94.0  20160513.0  routine  2016.0  
4 -122.421547  +14157763262   94.0  20141110.0  routine  2014.0  
5 -122.421547  +14157763262   94.0  20140214.0  routine  2014.0  
6 -122.403135  +14156779494   98.0  20161005.0  routine  2016.0  
7 -122.403135  +14156779494   96.0  20160311.0  routine  2016.0  
8 -122.403135  +14156779494   96.0  20141124.0  routine  2014.0  
9 -122.403135  +14156779494   96.0  20140612.0  routine  2014.0

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [161]:
full_table2 = full_table.sort_values(['score'], ascending=True)

full_table2[['name', 'address', 'score']].head(20)




name                     address  \
10664                        ABC BAKERY CAFE             650 JACKSON ST    
6362                         IMPERIAL PALACE          818 WASHINGTON ST    
9258          YUMMY DIM SUM & FAST FOOD, LLC            930 STOCKTON ST    
11708                GOLDEN RIVER RESTAURANT            5827 GEARY BLVD    
2446                             LES JOULINS                44 ELLIS ST    
2426                        YEE'S RESTAURANT             1131 GRANT AVE    
6710                     BRISTOL FARMS #2317           845 MARKET ST 010   
5907                           BODEGA BISTRO              607 LARKIN ST    
8629        OZONE THAI RESTAURANT AND LOUNGE                598 02ND ST    
2715             BASIL THAI RESTAURANT & BAR             1175 FOLSOM ST    
12076                        MISS SAIGON LLC                100 06TH ST    
11665                             UNCLE CAFE              65 WAVERLY PL    
14587                HARBOR VILLA RESTAURANT  5238 DIAMOND HEIGHTS BLVD    
7749                              FIRST CAKE            5901 GEARY BLVD    
15689                              POKI TIME            2101 LOMBARD ST    
14455                           CRAZY PEPPER          2257 SAN JOSE AVE    
11651                          SA BEANG THAI          312 DIVISADERO ST    
3362                         GOURMET KITCHEN           1051 STOCKTON ST    
14400  LE SOLEIL INTERNATIONAL HOLDINGS, INC          133 CLEMENT  ST.     
858                          TAQUERIA LA PAZ                305 TURK ST    

       score  
10664   46.0  
6362    47.0  
9258    50.0  
11708   52.0  
2446    53.0  
2426    53.0  
6710    54.0  
5907    54.0  
8629    54.0  
2715    54.0  
12076   54.0  
11665   55.0  
14587   55.0  
7749    55.0  
15689   55.0  
14455   55.0  
11651   55.0  
3362    55.0  
14400   55.0  
858     56.0

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!